alns/accept/NonLinearGreatDeluge.py

```diff
@@ -0,0 +1,106 @@
+import math
+from typing import Optional
+
+from alns.accept.GreatDeluge import GreatDeluge
+
+
+class NonLinearGreatDeluge(GreatDeluge):
+    """
+    The Non Linear Great Deluge (NLGD) criterion accepts solutions if the
+    candidate solution has value lower than a threshold (originally called the
+    water level [1]). The initial threshold is computed as
+
+    ``threshold = alpha * initial.objective()``
+
+    where ``initial`` is the initial solution passed-in to ALNS, inferred
+    from the best solution at the first iteration.
+
+    The non-linear GD variant was proposed by [2]. It differs from GD by using
+    a non-linear updating scheme, see the ``_update`` method for more details.
+    Moreover, candidate solutions that improve the current solution are always
+    accepted.
+
+    The implementation is based on the description in [2].
+
+    Parameters
+    ----------
+    alpha
+        Factor used to compute the initial threshold
+    beta
+        Factor used to update the threshold
+    gamma
+        Facto1 used to update the threshold
+    delta
+        Factor used to update the threshold
+
+    References
+    ----------
+    [1]: Dueck, G. New optimization heuristics: The great deluge algorithm and
+         the record-to-record travel. *Journal of Computational Physics* (1993)
+         104 (1): 86-92.
+    [2]: Landa-Silva, D., & Obit, J. H. Great deluge with non-linear decay rate
+         for solving course timetabling problems. *4th international IEEE
+         conference intelligent systems* (2008) Vol. 1: 8-11.
+    [3]: Santini, A., Ropke, S. & Hvattum, L.M. A comparison of acceptance
+         criteria for the adaptive large neighbourhood search metaheuristic.
+         *Journal of Heuristics* (2018) 24 (5): 783–815.
+    """
+
+    def __init__(self, alpha: float, beta: float, gamma: float, delta: float):
+        super().__init__(alpha, beta)
+
+        if gamma <= 0 or delta <= 0:
+            raise ValueError("Gamma and delta must be non-negative.")
+
+        self._gamma = gamma
+        self._delta = delta
+
+    @property
+    def gamma(self):
+        return self._gamma
+
+    @property
+    def delta(self):
+        return self._delta
+
+    def __call__(self, rnd, best, curr, cand):
+        if best.objective() == 0:
+            raise ValueError("Initial solution cannot have zero value.")
```

Comment:
Question: why not? What goes wrong?

Reply:
Long story: the initial threshold is computed using the initial solution * alpha, where the initial solution is inferred best in the the first call. 

If best is zero, then the initial threshold is zero. During the update, the rel gap between candidate and threshold has to be computed. But we get a ZeroDivisionError in this case.

Reply:
Ah OK, that makes sense.